### PR TITLE
Tidy flexlink bootstrap

### PR DESCRIPTION
**Pull request by dra27**
When bootstrapping flexlink, the usual invocation of `flexlink` (`FLEXLINK_CMD` in `config/Makefile`) has to be overridden as, like everything else during the compiler's build process, the command must be prefixed with an explicit call to `ocamlrun`. This is done in two places:
 1. `FLEXLINK_CMD` is threaded from the toplevel `Makefile`, which means that any `Makefile` commands which directly invoke `$(FLEXLINK)` automagically work.
 2. `OCAML_FLEXLINK` has to be exported to the environment when running `ocamlc -custom`, linking an executable with `ocamlopt -o`, and running `ocamlopt -shared` to instruct the compiler itself not to invoke `flexlink` directly.

Some `Makefile`s achieved the second by having an `export OCAML_FLEXLINK := ...` line and other more recent ones had a more elegant `FLEXLINK_ENV = OCAML_FLEXLINK=...` which is then explicitly used in commands when required. The latter has the benefit of appearing less magical, because you can see the `OCAML_FLEXLINK="../boot/ocamlrun ../flexdll/flexlink.exe"` instruction on the individual command lines and so be certain that the correct flexlink is being called.

This GPR, in preparation for full support of the Windows Linux Subsystem for building OCaml, tidies up three things:
 1. `FLEXLINK_ENV` is now used everywhere, instead of an exported variable
 2. The definition of `FLEXLINK_ENV` is moved to `Makefile.common` and consequentially several `Makefile`s now define and consistently use `ROOTDIR`
 3. `FLEXLINK_ENV` is used exactly where it is needed - previously it was used on all `CAMLOPT` invocations, which is unnecessary.

### CODE REVIEW

**Comment by dra27** [comment]
I've put this branch through [precheck](https://ci.inria.fr/ocaml/job/precheck/135/) as well.

**Comment by shindere** [comment]
Yep, just reviewing this before leaving for a two weeks. :)

For adding a space at the end of `FLEXLINK_ENV`, wouldn't it be better
to use the `$()EMPTY` trick as is already done for `OUTPUTOBJ` ?
Or shall we do things the other way around and use a comment also for
`OUTPUTOBJ` ?

And by teh way why must the space be at the end of `FLEXLINK_ENV`,
actually? Why is it necessary to write

```shell
+       $(FLEXLINK_ENV)$(CAMLOPT) $(LINKFLAGS) -o $@ $^ -cclib "$(BYTECCLIBS)"
```

rather than

```shell
+       $(FLEXLINK_ENV) $(CAMLOPT) $(LINKFLAGS) -o $@ $^ -cclib "$(BYTECCLIBS)"
```

In the modification of ocamltest's Makefile to use `ROOTDIR`, perhaps
you coud use the make's `addprefix` function to make the definition of
the `directories` variable shorter?

Just out of curiosity: the definitions of `FLEXLINK_ENV` seem to
consistently use spaces around the `=` sign between FLEXLINK_ENV and its
definition, while in ROOTDIR this does not seem to be the case. Would
you mind putting spaces everywhere? And also, would you mind using
`:=`rather than `=` when possible?

All this being said: I find this PR really cool and am going to approve
it to make sure my vacations do not block it. Thanks a lot, @dra27!

**Review by shindere — APPROVED** [review]

**Comment by dra27** [comment]
Thanks for the review, @shindere (and have a good vacation - I too am on vacation next week!).

I debated doing the `$(EMPTY)` trick for the space, but I kept hitting line length issues, which was my only reason for using end-of-line commenting instead! I get a bit neurotic about stray spaces in commands - if you have `$(FLEXLINK_ENV) $(CAMLOPT)` then you would see a stray space at the start of the console output when `FLEXLINK_ENV` is empty (i.e. most of the time). I get similarly neurotic elsewhere in the build system where you get lots of spaces in a row, particularly in the stdlib!

As a rule of thumb, I only use `:=` when there are side-effects like `$(shell` which I definitely only want run once or when it's updating a variable which has been derived iteratively (e.g. `CAMLOPT := blah $(CAMLOPT)`). Here I think it's a bad idea - although I have ensure that `ROOTDIR` is defined before `Makefile.common` is `include`d, defining `FLEXLINK_ENV` with `=` means that you don't **have** to define `ROOTDIR` before `include`ing `Makefile.common` - with `:=` you must define it first.

**Comment by dra27** [comment]
(actually, having just gone through the changes again to put the spaces and much neater `$(addprefix` calls in, `ROOTDIR` does still have to be defined before `Makefile.common` is included because the test for the submodule is done using `:=`)

**Comment by shindere** [comment]
David Allsopp (2018/07/27 03:18 -0700):
> Thanks for the review, @shindere

My pleasure! It's great to see all theese cleanups on the build system!

> (and have a good vacation - I too am on vacation next week!).

Enjoy!

> I debated doing the `$(EMPTY)` trick for the space, but I kept hitting
> line length issues, which was my only reason for using end-of-line
> commenting instead!

Understood, thanks a lot for clarifying.

> I get a bit neurotic about stray spaces in 
> commands - if you have `$(FLEXLINK_ENV) $(CAMLOPT)` then you would see
> a stray space at the start of the console output when `FLEXLINK_ENV`
> is empty (i.e. most of the time).

Ah, I understand. I tend to dislike extraneous spaces, too, but at the
same time it seems to me that putting the space at the end of the
variable rather than between the variable and the command makes the
system a bit fragile. I mean: if, for some reason, the space is
forgotten, that will result in an invalid command. Whereas if the space
is in the command, then the command may of course contain a useless
space but would be valid by construction and it would'nt be possible
(or, it wouldn't be as easy) to break it just by misdefining
FLEXLINK_ENV. So I think, all in all, I'd prefer the space in the
command but at the same time I have no strong opinion on it so if you
wish to keep the code as it is I'm fine with that and we can always
change this later.

> I get similarly neurotic elsewhere
> in the build system where you get lots of spaces in a row,
> particularly in the stdlib!

I know ocamltest does that, too, and that's definitely
something I'd like to improve int the future because I see these things
(at least in the code I write myself) as symptoms of not-so-well written
code.

> As a rule of thumb, I only use `:=` when there are side-effects like
> `$(shell` which I definitely only want run once or when it's updating
> a variable which has been derived iteratively (e.g. `CAMLOPT := blah
> $(CAMLOPT)`). Here I think it's a bad idea - although I have ensure
> that `ROOTDIR` is defined before `Makefile.common` is `include`d,
> defining `FLEXLINK_ENV` with `=` means that you don't **have** to
> define `ROOTDIR` before `include`ing `Makefile.common` - with `:=` you
> must define it first.

Yeah and that's precisely what I like because, to me, it results in
makefiles that are easier to read and understand. But again, that's not
a blocker to me so, as far as I am concerned, please feel free to mergge
whenever you are yourself happy with the code.

Oh and by the way: given that users may actually want to bootstrap
flexdll, wouldn't this PR diserve a Changes entry?

**Comment by shindere** [comment]
@dra27: do you think we could get this merged soon? Because your cleanups
would actually also be useful for the work on autoconf.

Coming back to the spaces between FLEXLINK_ENV and commands thing,
I might have another proposal that seems satisfacotry to me.

How about defining variables like

```shell
CAMLOPT_CMD = $(if $(FLEXLINK_ENV), $(FLEXLINK_ENV) $(CAMLOPT), $(CAMLOPT))
```


and then invoking `$(CAMLOPT_CMD)` rather than `$(CAMLOPT)` ?

Would that seem too heavy ?

**Comment by dra27** [comment]
@shindere - what do you think of 073111a? I'm not overly keen on it:
 - `FLEXLINK_ENV` is now permanently defined only in the two places which need it (`Makefile.common` and `testsuite/Makefile`), so the only space-related danger would be a new rule which *uses* `FLEXLINK_ENV` and adds an extraneous space after it (it's cosmetic of course, not a danger).
 - With the new version, the difference between `$(CAMLOPT)` and `$(CAMLOPT_CMD)` is not as clear as it was before. Before, it was very clear that you were setting the flexlink variable and that the command needed it. That's not brittle - if you don't define it when needed, then the bootstrapped build will fail on AppVeyor, so it's just a matter of clarity. It would also be possible to define `OCAML_FLEXLINK` all the time, I just have a preference for defining things when they are actually needed (I have the same opinion with the myriad `-I` options in the build system, but one battle at a time......!).

Anyway, this is rebased and ready to merge with or without the final commit (I don't mind!).

**Comment by shindere** [comment]
Many thanks for having taken the time to do that, David!

Yes, I have to say I like it with the last commit so let me merge with
it and we can always change things again later if there is a wish to do
so!

**Review by shindere — APPROVED** [review]

**Comment by shindere** [comment]
@dra27: I can't see what wrong with Travis, do you know, please?
Do we need to wait until it passes to merge?

Same for AppVeyor which is still pending...?

**Comment by dra27** [comment]
It has passed on my AppVeyor, but we should still wait.

**Comment by dra27** [comment]
Oops - I think I've broken something, looking at Travis 😊

**Comment by shindere** [comment]
@dra27 feel free to merge yourself when you are happy with it. I mean I
can do it tomorrow if everything is okay and it has not been merged, but
no need to wait for me if it's okay for you to merge yourself.

**Comment by dra27** [comment]
So, the Changes job is failing because of GitHub API limits and the AppVeyor mingw32 testsuite ran out of time, but the msvc64 testsuite passed. So I'll take a risk and merge this!

Thanks for the reviews @shindere!

**Comment by shindere** [comment]
David Allsopp (2018/09/06 23:34 -0700):
> So, the Changes job is failing because of GitHub API limits and the
> AppVeyor mingw32 testsuite ran out of time, but the msvc64 testsuite
> passed. So I'll take a risk and merge this!

Fully agreed! Thanks!

If Inria's CI reports something I'll investigate and either fix myself
or report here if I can't.

> Thanks for the reviews @shindere!

My pleasure! Thanks to you for all this great work and your patience and
flexibility!

**Comment by gasche** [comment]
In a [CI testsuite run](https://ci.inria.fr/ocaml/job/main/flambda=true,label=ocaml-macos-10.13/1155/console) on MacOS (failed for some other reason), I see many warnings of the form

```

Running tests from 'tests/lib-queue' ...
../Makefile.tools:53: warning: undefined variable `USE_RUNTIME'
../Makefile.tools:77: warning: undefined variable `FLEXLINK'
Makefile:54: warning: undefined variable `FLEXLINK_ENV'
 ... testing 'test.ml' with 1 (native) => passed
 ... testing 'test.ml' with 2 (bytecode) => passed
Running tests from 'tests/lib-random' ...
../Makefile.tools:53: warning: undefined variable `USE_RUNTIME'
../Makefile.tools:77: warning: undefined variable `FLEXLINK'
Makefile:54: warning: undefined variable `FLEXLINK_ENV'
 ... testing 'rand.ml' with 1 (native) => passed
 ... testing 'rand.ml' with 2 (bytecode) => passed
Running tests from 'tests/lib-result' ...
../Makefile.tools:53: warning: undefined variable `USE_RUNTIME'
../Makefile.tools:77: warning: undefined variable `FLEXLINK'
Makefile:54: warning: undefined variable `FLEXLINK_ENV'
 ... testing 'test.ml' with 1 (native) => passed
 ... testing 'test.ml' with 2 (bytecode) => passed
```

Are those caused by this PR? Should testsuite/makefiles/Makefile.common include a definition of these variables somehow?

**Comment by shindere** [comment]
NO, the warnings are not caused by this PR.

I have a branch to fix this, just need to write the PR text. Will do on
Monday.

**Pull request merged**
